# Exploring Hacker News Posts

[Hacker News](https://news.ycombinator.com/) is a site started by the startup incubator [Y Combinator](https://www.ycombinator.com/), where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.

In this Project, a data set spanning 12 months will be analyzed.

The data set has been reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions. Below are descriptions of the columns:

- id: The unique identifier from Hacker News for the post
- title: The title of the post
- url: The URL that the posts links to, if it the post has a URL
- num_points: The number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes
- num_comments: The number of comments that were made on the post
- author: The username of the person who submitted the post
- created_at: The date and time at which the post was submitted

We're specifically interested in posts whose titles begin with either Ask HN or `Show HN`. Users submit `Ask HN` posts to ask the Hacker News community a specific question. Example:
```
Ask HN: How to improve my personal website?
Ask HN: Am I the only one outraged by Twitter shutting down share counts?
Ask HN: Aby recent changes to CSS that broke mobile?
```

Likewise, users submit `Show HN` posts to show the Hacker News community a project, product, or just generally something interesting. Example:
```
Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform'
Show HN: Something pointless I made
Show HN: Shanhu.io, a programming playground powered by e8vm
```

We'll compare these two types of posts to determine the following:
- Do Ask HN or Show HN receive more comments on average?
- Do posts created at a certain time receive more comments on average?

### Let's start by importing the libraries we need and reading the data set into a list of lists.

In [1]:
from csv import reader
open_file = open('hacker_news.csv')
read_file = reader(open_file)
hn = list(read_file)

print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


In [2]:
headers = hn[:1] #extract header
hn = hn[1:]      #delete header in hn list
print('Total number of posts:',len(hn),'\n')
print(headers, '\n')
print('-' * 75,'\n')
for post in hn[0:5]:
    print(post,'\n')

Total number of posts: 20100 

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']] 

--------------------------------------------------------------------------- 

['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'] 

['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'] 

['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'] 

['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'] 

['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http

# Data filtering

As explained in the beginning of the project, we're only concerned with post titles beginning with `Ask HN` or `Show HN`

In the section below, we will seperate the posts in the data set on that basis.

- `ask_posts`: List of data containing only the Ask HN posts
- `show_posts`: List of data containing only the Show HN posts
- `other_posts`: List of data containing posts other than the above two



In [3]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    title = title.lower()
    if title.startswith('ask hn'):
        ask_posts.append(row)
    elif title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)

print('Number of Ask HN posts:',len(ask_posts))
print('Number of Show HN posts:',len(show_posts))
print('Number of Other HN posts:',len(other_posts))

Number of Ask HN posts: 1744
Number of Show HN posts: 1162
Number of Other HN posts: 17194


# Data analysis
### Step 1: Comments - Ask HN vs Show HN posts

In [4]:
total_ask_comments = 0
for row in ask_posts:
    num_comments = int(row[4])
    total_ask_comments += num_comments
print('Total ask comments:',total_ask_comments)
avg_ask_comments = total_ask_comments / len(ask_posts)
print('Average number of comments on ask posts',avg_ask_comments)

total_show_comments = 0
for row in show_posts:
    num_comments = int(row[4])
    total_show_comments += num_comments
print('Total show comments:',total_show_comments)
avg_show_comments = total_show_comments / len(show_posts)
print('Average number of comments on show posts',avg_show_comments)
ratio_ask_show_posts = len(ask_posts) / len(show_posts)
print('Ratio of ask posts to show posts',ratio_ask_show_posts)

Total ask comments: 24483
Average number of comments on ask posts 14.038417431192661
Total show comments: 11988
Average number of comments on show posts 10.31669535283993
Ratio of ask posts to show posts 1.5008605851979346


##### Discussion
> Even though there are about 1.5 as many ask posts than show posts, on average the ask posts receive more comments than the show posts.
This can be attributed to the much higher number of total ask comments (24,483) compared to the show comments (11,988).

### Step 2: Analyse the number of posts and total comments by the hour created
We will now see the number of posts made and total number of comments received based on the hour of the day the post was created (Based on the `created_at column`).

**Note:** According to the data set documentation, the timezone used is Eastern Time in the US.

Next, we'll determine if ask posts created at a certain time are more likely to attract comments. We'll use the following steps to perform this analysis:

1. Calculate the amount of ask posts created in each hour of the day, along with the number of comments received.
2. Calculate the average number of comments ask posts receive by hour created.

Let's calculate the amount of ask posts created per hour, along with the total amount of comments first:

In [5]:
import datetime as dt

result_list = []

for row in ask_posts:
    created_at = row[6]
    num_comments = int(row[4])
    result_list.append([created_at, num_comments])

counts_by_hour = {}
comments_by_hour = {}

for row in result_list:
    date_time = row[0]
    num_comments = row[1]
    dt_object = dt.datetime.strptime(date_time, '%m/%d/%Y %H:%M')
    #hour = dt_object.hour #this method would extract the hour not as string
    hour = dt_object.strftime("%H")
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = row[1]
    else:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += row[1]
        
print('Amount of ask posts created each hour:','\n',counts_by_hour,'\n')
print('Amount of comments created each hour:','\n',comments_by_hour)

Amount of ask posts created each hour: 
 {'16': 108, '23': 68, '20': 80, '15': 116, '06': 44, '00': 55, '13': 85, '18': 109, '05': 46, '09': 45, '02': 58, '14': 107, '01': 60, '04': 47, '07': 34, '03': 54, '10': 59, '08': 48, '11': 58, '17': 100, '12': 73, '21': 109, '19': 110, '22': 71} 

Amount of comments created each hour: 
 {'16': 1814, '23': 543, '20': 1722, '15': 4477, '06': 397, '00': 447, '13': 1253, '18': 1439, '05': 464, '09': 251, '02': 1381, '14': 1416, '01': 683, '04': 337, '07': 267, '03': 421, '10': 793, '08': 492, '11': 641, '17': 1146, '12': 687, '21': 1745, '19': 1188, '22': 479}


Above, we created two dictionaries and displayed them:

- `counts_by_hour`: contains the number of ask posts created during each hour of the day.
- `comments_by_hour`: contains the corresponding number of comments ask posts created at each hour received.

Right now, we'll use those two dictionaries to calculate the average number of comments for posts created during each hour of the day.

In [6]:
avg_by_hour = []

for row in comments_by_hour:
    avg_by_hour.append([row, comments_by_hour[row] / counts_by_hour[row]])
    
print(avg_by_hour)

[['16', 16.796296296296298], ['23', 7.985294117647059], ['20', 21.525], ['15', 38.5948275862069], ['06', 9.022727272727273], ['00', 8.127272727272727], ['13', 14.741176470588234], ['18', 13.20183486238532], ['05', 10.08695652173913], ['09', 5.5777777777777775], ['02', 23.810344827586206], ['14', 13.233644859813085], ['01', 11.383333333333333], ['04', 7.170212765957447], ['07', 7.852941176470588], ['03', 7.796296296296297], ['10', 13.440677966101696], ['08', 10.25], ['11', 11.051724137931034], ['17', 11.46], ['12', 9.41095890410959], ['21', 16.009174311926607], ['19', 10.8], ['22', 6.746478873239437]]


Although we now have the results we need, this format makes it hard to identify the hours with the highest values. Let's finish by sorting the list of lists and printing the five highest values in a format that's easier to read.

In [12]:
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])

#print(swap_avg_by_hour)

sorted_swap = sorted(swap_avg_by_hour, reverse=True)
print(sorted_swap)

[[38.5948275862069, '15'], [23.810344827586206, '02'], [21.525, '20'], [16.796296296296298, '16'], [16.009174311926607, '21'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [13.20183486238532, '18'], [11.46, '17'], [11.383333333333333, '01'], [11.051724137931034, '11'], [10.8, '19'], [10.25, '08'], [10.08695652173913, '05'], [9.41095890410959, '12'], [9.022727272727273, '06'], [8.127272727272727, '00'], [7.985294117647059, '23'], [7.852941176470588, '07'], [7.796296296296297, '03'], [7.170212765957447, '04'], [6.746478873239437, '22'], [5.5777777777777775, '09']]


In [17]:
print('Top 5 Hours for Ask Posts Comments')
#for average, hour in sorted_swap[:5]:
#    output = '{h}:00: {a:.2f} average comments per post'.format(h=hour, a=average)
#    print(output)
hour_format = "%H:%M"    # the format we want the time to take
for row in sorted_swap[:5]:
    avg = row[0]
    hour = str(row[1])   # converts the hours to string format
    
    hour = dt.datetime.strptime(hour, "%H")  # parses the hours as a datetime object
    hour = hour.strftime(hour_format)        # specifies the format of the time
    
    # uses the format to print the hour and average
    hour_avg_string = "{h}: {a:.2f} average comments per posts".format(h=hour, a=avg)
    print(hour_avg_string)   # displays the top 5 hours according to the average comments

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per posts
02:00: 23.81 average comments per posts
20:00: 21.52 average comments per posts
16:00: 16.80 average comments per posts
21:00: 16.01 average comments per posts


**Note:** According to the data set documentation, the timezone used is Eastern Time in the US.

Therefore, if one lives in the **Eastern Time Zone in the US**, one should post around 15:00 to have a higher chance of receiving comments.

If one lives in **Vienna, Austria**, one should therefore post around 21:00 to have a higher chance of receiving comments.



# Conclusion

Our primary aim in this analysis was to determine the best time to send posts on the [Hacker News platform](https://news.ycombinator.com/). This became necessary because most users become frustrated and disappointed when their questions go unanswered. We believe the understanding of a platform and the activity of its users is paramount to the attraction engagement/attention to a post.

Therefore, we analyzed the average comments received by posts sent within each hour to determine the hours that has the most comments per post. In our result we found that:

- Ask posts (i.e. posts that come in form of questions) generally has more average number of comments more that Show posts.
- Posts sent within the hour of 15:00 (US Eastern Time) has the most engagement in form of comments.

We can therefore suggest that users should send their posts within this hour in order for their posts to receive highest attention.


### Limitations

In this analysis, we did not:

- determine the hours the comments received by the posts were sent (this might be necessary to understand the time interval between the time the posts were created and the time the comments were received)
- determine the average number of points received by the posts
